ls1intum/Artemis
# Feature/Track user interactions in assessment of text exercise

### Checklist


### Motivation and Context


Resolves #3649 
> Improvements on the assesment system and its infrastructure are intended to make grading easier and more efficient. However, it is still not clear whether certain improvements prove to be efficient/useful or not. Mainly, we find it that in practice, some added functionality, is sometimes seldom used (i.e.: Conclict Resolution). Sooner than later, instead of having a simple to use and efficient grading process, we end up saturating it with endless features, without knowing whether they are currently being used actively or not.

_The aim of this PR is to provide a way to properly identify usage analytics of features in the grading process of text exercises._ 


### Description

- Added a new Table, with its respective Entity, Resource and Repository:
![image](https://user-images.githubusercontent.com/18335489/125179096-597f2100-e1eb-11eb-9ddc-c3b696c72543.png)
- Added a toggle functionality to switch this feature on/off. This is based upon the `info.text-assessment-analytics-enabled: true` property used in the **.yml** files as an easier way to apply in production. This value is checked both in the server side resource as well as within the `TextAssessmentAnalytics` service on the client side, where we fetch this value from the server through the ProfileService. 
- Added a client side service: `TextAssessmentAnalytics` that can be injected in any component and used to easily send events to the server. All http calls are **non-blocking** on the client side. No **await** is used to avoid button clicks being delayed. Also the response from the server is not something we are interested in checking back. 
- Added **new** Server and Java tests for newly added functionality

### Steps for Testing


1. Log in to Artemis
2. Navigate to Course Administration
3. Open a Text Exercise submission.
       - Alternatively access different submissions through the links below: (test user 11-15)
       - [Some list of ready submissions](https://artemistest.ase.in.tum.de/course-management/482/text-exercises/7729/submissions)
       - [Submission with a Conflict](https://artemistest.ase.in.tum.de/course-management/482/text-exercises/7750/participations/323633/submissions/438009/assessment)
        - [Simple submission feedback](https://artemistest.ase.in.tum.de/course-management/482/text-exercises/7832/participations/323822/submissions/438207/assessment)
       - [Submission without feedback](https://artemistest.ase.in.tum.de/course-management/482/text-exercises/7733/participations/323611/submissions/437980/assessment)
       - Also check that it works the same with **Exam Text Exercises**: (Example Exam below)
       - [Link to Exam Text Exercise Submission](https://artemistest.ase.in.tum.de/course-management/482/exams/1012/exercise-groups/2226/text-exercises/9807/participations/325301/submissions/439848/assessment?correction-round=0)
4. Interact with the Text Assessment Area elements as shown in the screenshot attached
5. Each click/interaction with the elements should make an **add-event** call. Check your Developer Tool/Network
6. Make sure the calls respond with **200** and the body is not malformed
7. Make sure to test for at least a **Teaching Assistant**, TA, that the event are sent from them as well. (test user 6-9)

### Test Coverage



| File                                                                                                  | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s    
| assessment-header.component.ts                                                  |     100 |    85.71 |     100 |     100 | 33-56|   
| assessment-layout.component.ts                                                   |     100 |       75 |     100 |     100 | 39-43|
| textblock-assessment-card.component.ts                                     |    87.1 |    72.22 |   71.43 |   87.04 | 98-107 |
textblock-feedback-editor.component.ts                                           |    72.9 |    58.59 |   57.69 |   72.34 | 81,91-92,107,122-125,143-166,197-217|
| manual-text-selection.component.ts                                               |     100 |    78.57 |     100 |     100 | 23,57 |
| text-assesment-analytics.service.ts                                              |   70.97 |       75 |   42.86 |   66.67 | 34,62-79 |

### Setup Documentation
![image](https://user-images.githubusercontent.com/18335489/125362194-c7078a80-e36e-11eb-9866-7a70edc5c060.png)


### Screenshots


![psd_file](https://user-images.githubusercontent.com/18335489/123979287-e1318800-d9c0-11eb-99f9-0434f94c4bce.png)

![c010](https://user-images.githubusercontent.com/18335489/125179585-df9d6680-e1ef-11eb-8d69-ac8eea4b9b59.PNG)
